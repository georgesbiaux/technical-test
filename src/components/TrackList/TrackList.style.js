import styled from 'styled-components';

const StyledTrackList = styled.div`
  position: relative;

  .track-list-transition-group {
    position: relative;
  }

  .fade-enter {
    opacity: 0.01;
  }

  .fade-enter.fade-enter-active {
    opacity: 1;
    transition: opacity 250ms ease-in;
  }

  .fade-exit {
    opacity: 1;
  }

  .fade-exit.fade-exit-active {
    opacity: 0.01;
    transition: opacity 200ms ease-in;
  }

  .track-item {
    position: absolute;
    right: 0;
    left: 0;
    transition: top 500ms;
  }
`;

export default StyledTrackList;
