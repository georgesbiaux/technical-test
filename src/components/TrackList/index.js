export { default } from './TrackList.container';
