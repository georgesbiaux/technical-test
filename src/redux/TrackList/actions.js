import {
  TOGGLE_TRACK_PRIORITY,
  TOGGLE_USER_VOTE,
  ADD_TRACK,
} from './constants';

export function toggleTrackPriority(trackId) {
  return {
    type: TOGGLE_TRACK_PRIORITY,
    trackId,
  };
}

export function toggleUserVote(trackId) {
  return {
    type: TOGGLE_USER_VOTE,
    trackId,
  };
}

export function addTrack(track) {
  return {
    type: ADD_TRACK,
    track,
  };
}
