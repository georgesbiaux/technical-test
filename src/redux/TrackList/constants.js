export const TOGGLE_TRACK_PRIORITY = 'TOGGLE_TRACK_PRIORITY';
export const TOGGLE_USER_VOTE = 'TOGGLE_USER_VOTE';
