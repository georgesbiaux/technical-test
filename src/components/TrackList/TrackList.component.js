import React from 'react';
import { TrackItem, EmptyTrackList } from 'components';
import StyledTrackList from './TrackList.style';

const TrackList = props => {
  let timeBeforePlay = 240;
  const trackList = props.trackList.map(track => {
    const item = (
      <TrackItem key={track.id} track={track} timeBeforePlay={timeBeforePlay} />
    );

    timeBeforePlay += track.duration;
    return item;
  });

  return (
    <StyledTrackList className="track-list">
      {props.trackList.length > 0 ? trackList : <EmptyTrackList />}
    </StyledTrackList>
  );
};

export default TrackList;
