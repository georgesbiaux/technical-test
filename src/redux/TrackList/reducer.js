// @flow

import { fromJS } from 'immutable';
import { TOGGLE_TRACK_PRIORITY, TOGGLE_USER_VOTE } from './constants';

const initialState = fromJS({
  tracks: [
    {
      id: 1,
      name: 'Beast Of Burden',
      duration: 211,
      priority: false,
      artist: 'The Rolling Stones',
      adder: {
        id: 1,
        name: 'kant',
        pictureUrl:
          'https://res.cloudinary.com/jukeo-net/image/upload/ano-b2_eezggd',
      },
      pictureUrl:
        'https://i.scdn.co/image/adcf79a6adae4e1a44d0e34860cc103e28ee4372',
      votes: {
        count: 0,
        userVoted: false,
      },
    },
    {
      id: 2,
      name: 'Purple Haze',
      duration: 229,
      priority: false,
      artist: 'Jimi Hendrix',
      adder: {
        id: 1,
        name: 'kant',
        pictureUrl:
          'https://res.cloudinary.com/jukeo-net/image/upload/ano-b2_eezggd',
      },
      pictureUrl:
        'https://i.scdn.co/image/1d55f445789ed89a1b8fab09c3dad117afb5fe80',
      votes: {
        count: 0,
        userVoted: false,
      },
    },
  ],
});

export default function reducer(
  state: Map<string, any> = initialState,
  action: { type: string, payload: Object },
) {
  switch (action.type) {
    case TOGGLE_TRACK_PRIORITY: {
      const tracks = state.get('tracks').toJS();
      const trackIndex = tracks.findIndex(track => track.id === action.trackId);

      tracks[trackIndex].priority = !tracks[trackIndex].priority;

      return state.set('tracks', fromJS(tracks));
    }
    case TOGGLE_USER_VOTE: {
      const tracks = state.get('tracks').toJS();
      const trackIndex = tracks.findIndex(track => track.id === action.trackId);

      if (tracks[trackIndex].votes.count === 0) {
        tracks[trackIndex].votes.userVoted = true;
      }

      tracks[trackIndex].votes.count += 1;

      return state.set('tracks', fromJS(tracks));
    }
    default:
      return state;
  }
}
