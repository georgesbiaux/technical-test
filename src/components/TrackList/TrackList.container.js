import { connect } from 'react-redux';
import { toJS } from '../../services/immutable/toJs';
import TrackList from './TrackList.component';

function mapStateToProps() {
  return {};
}

function mapDispatchToProps() {
  return {};
}

export default connect(mapStateToProps, mapDispatchToProps)(toJS(TrackList));
