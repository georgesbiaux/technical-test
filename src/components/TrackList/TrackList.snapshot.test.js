import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import TrackList from './TrackList.component';

describe('<TrackList>', () => {
  it('should render correctly', () => {
    const props = {
      trackList: [{
        id: 1,
        duration: 500,
      }, {
        id: 2,
      }],
    };
    const tree = shallow(<TrackList {...props} />);
    expect(toJSON(tree)).toMatchSnapshot();
  });

  it('should render correctly when tracklist is empty', () => {
    const props = {
      trackList: [],
    };
    const tree = shallow(<TrackList {...props} />);
    expect(toJSON(tree)).toMatchSnapshot();
  });
});
