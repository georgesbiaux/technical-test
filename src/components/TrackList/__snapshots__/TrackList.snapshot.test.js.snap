// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackList> should render correctly 1`] = `
<styled.div
  className="track-list"
>
  <TrackItem
    key="1"
    timeBeforePlay={240}
    track={
      Object {
        "duration": 500,
        "id": 1,
      }
    }
  />
  <TrackItem
    key="2"
    timeBeforePlay={740}
    track={
      Object {
        "id": 2,
      }
    }
  />
</styled.div>
`;
